cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -448,5 +631,320 @@ sample_edges(
+           std::optional<rmm::device_uvector<int32_t>>,
+           std::optional<rmm::device_uvector<label_t>>>
+shuffle_sampling_results(raft::handle_t const& handle,
+                         rmm::device_uvector<vertex_t>&& majors,
+                         rmm::device_uvector<vertex_t>&& minors,
+                         std::optional<rmm::device_uvector<weight_t>>&& weights,
+                         std::optional<rmm::device_uvector<edge_t>>&& edge_ids,
+                         std::optional<rmm::device_uvector<edge_type_t>>&& edge_types,
+                         std::optional<rmm::device_uvector<int32_t>>&& hops,
+                         std::optional<rmm::device_uvector<label_t>>&& labels,
+                         raft::device_span<int32_t const> label_to_output_gpu_mapping)
+{
+  CUGRAPH_EXPECTS(labels, "labels must be specified in order to shuffle sampling results");
+
+  auto& comm           = handle.get_comms();
+  auto const comm_size = comm.get_size();
+
+  auto total_global_mem = handle.get_device_properties().totalGlobalMem;
+  auto element_size     = sizeof(vertex_t) * 2 + (weights ? sizeof(weight_t) : size_t{0}) +
+                      (edge_ids ? sizeof(edge_t) : size_t{0}) +
+                      (edge_types ? sizeof(edge_type_t) : size_t{0}) +
+                      (hops ? sizeof(int32_t) : size_t{0}) + sizeof(label_t);
+
+  auto constexpr mem_frugal_ratio =
+    0.1;  // if the expected temporary buffer size exceeds the mem_frugal_ratio of the
+          // total_global_mem, switch to the memory frugal approach (thrust::sort is used to
+          // group-by by default, and thrust::sort requires temporary buffer comparable to the input
+          // data size)
+  auto mem_frugal_threshold =
+    static_cast<size_t>(static_cast<double>(total_global_mem / element_size) * mem_frugal_ratio);
+
+  auto mem_frugal_flag =
+    host_scalar_allreduce(comm,
+                          majors.size() > mem_frugal_threshold ? int{1} : int{0},
+                          raft::comms::op_t::MAX,
+                          handle.get_stream());
+
+  // invoke groupby_and_count and shuffle values to pass mem_frugal_threshold instead of directly
+  // calling groupby_gpu_id_and_shuffle_values there is no benefit in reducing peak memory as we
+  // need to allocate a receive buffer anyways) but this reduces the maximum memory allocation size
+  // by half or more (thrust::sort used inside the groupby_and_count allocates the entire temporary
+  // buffer in a single chunk, and the pool allocator  often cannot handle a large single allocation
+  // (due to fragmentation) even when the remaining free memory in aggregate is significantly larger
+  // than the requested size).
+  auto d_tx_value_counts = cugraph::groupby_and_count(
+    labels->begin(),
+    labels->end(),
+    [label_to_output_gpu_mapping] __device__(auto val) { return label_to_output_gpu_mapping[val]; },
+    comm_size,
+    mem_frugal_threshold,
+    handle.get_stream());
```

Comment:
You should shuffle (majors, minors, ...) with labels. I guess this is a bug here.

Reply:
Have you tested this in multi-GPU? If yes, we should update our test suites to detect this.